File: ASSIGNMENT3/EmployeeManagementSystem/src/InstructorEmployee.java
import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import exceptions.*;

public class InstructorEmployee extends Employee {

    private List<String> coursesTaught = new ArrayList<>();
    private String specialization;

    public InstructorEmployee(String firstName, String lastName, Date dateOfBirth,
            String employeeID, Position position, Manager manager,
            EmploymentType emplType, String specialization) {
        super(firstName, lastName, dateOfBirth, employeeID, position, manager, emplType);
        
        try {
            if (specialization == null || specialization.trim().isEmpty()) {
                throw new InvalidExceptions("Specialization cannot be empty");
            }
            this.specialization = specialization;
            
        } catch (InvalidExceptions e) {
            System.err.println("Error: " + e.getMessage());
            this.specialization = "General";
        }
    }

    public void addCourse(String course) {
        try {
            if (course == null || course.trim().isEmpty()) {
                throw new InvalidExceptions("Course name cannot be empty");
            }
            coursesTaught.add(course);
            
        } catch (InvalidExceptions e) {
            System.err.println("Error: " + e.getMessage());
        }
    }

    public List<String> getCoursesTaught() {
        return new ArrayList<>(coursesTaught);
    }

    public String getSpecialization() {
        return specialization;
    }

    public void setSpecialization(String specialization) {
        try {
            if (specialization == null || specialization.trim().isEmpty()) {
                throw new InvalidExceptions("Specialization cannot be empty");
            }
            this.specialization = specialization;
            
        } catch (InvalidExceptions e) {
            System.err.println("Error: " + e.getMessage());
        }
    }
}

File: Advanced_Prog_Assignment3-csharp/EmployeeManagementSystem/Models/InstructorEmployee.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using EmployeeManagementSystemCSharp.Exceptions;
using EmployeeManagementSystemCSharp.EmploymentTypes;

namespace EmployeeManagementSystemCSharp.Models
{
    public class InstructorEmployee : Employee
    {
        private readonly List<string> _coursesTaught = new List<string>();
        private string _specialization;

        public InstructorEmployee(string firstName, string lastName, DateTime dateOfBirth,
                                  string employeeID, Position position, Manager manager,
                                  EmploymentType emplType, string specialization)
            : base(firstName, lastName, dateOfBirth, employeeID, position, manager, emplType)
        {
            try
            {
                if (string.IsNullOrWhiteSpace(specialization))
                {
                    throw new InvalidExceptions("Specialization cannot be empty");
                }
                _specialization = specialization;
            }
            catch (InvalidExceptions e)
            {
                System.Console.Error.WriteLine("Error: " + e.Message);
            }
        }

        public void addCourse(string course)
        {
            try
            {
                if (string.IsNullOrWhiteSpace(course))
                {
                    throw new InvalidExceptions("Course name cannot be empty");
                }
                _coursesTaught.Add(course);
            }
            catch (InvalidExceptions e)
            {
                System.Console.Error.WriteLine("Error: " + e.Message);
            }
        }

        public List<string> getCoursesTaught()
        {
            return new List<string>(_coursesTaught);
        }

        public string getSpecialization()
        {
            return _specialization;
        }

        public void setSpecialization(string specialization)
        {
            try
            {
                if (string.IsNullOrWhiteSpace(specialization))
                {
                    throw new InvalidExceptions("Specialization cannot be empty");
                }
                _specialization = specialization;
            }
            catch (InvalidExceptions e)
            {
                System.Console.Error.WriteLine("Error: " + e.Message);
            }
        }

        public override void reportToManager(string message)
        {
            System.Console.WriteLine(
                $"Instructor {getFirstName()} {getLastName()} reports to manager: {message}");
        }
    }
}
